xc: handle kevent syscall interruption

daniloegea opened the pull request:
kevent will be interrupted and return EINTR on the arrival of a signal. When calling truss for example against the xcd child process that is handling the container it will crash because the kevent_ts will return Err and unwrap() is immediately called on it.

Check if kevent_ts failed and restart the loop if it failed due to a EINTR, panic! otherwise.

daniloegea committed:
xc: handle kevent syscall interruption

kevent will be interrupted and return EINTR on the arrival of a signal.
When calling truss for example against the xcd child process that is
handling the container it will crash because the kevent_ts will return
Err and unwrap() is immediately called on it.

Check if kevent_ts failed and restart the loop if it failed due to a
EINTR, panic! otherwise.

daniloegea commented on xc/src/container/runner/mod.rs:
```diff
@@ -600,8 +600,17 @@ impl ProcessRunner {
+                    if value == nix::errno::Errno::EINTR {
+                        continue;
+                    } else {
+                        panic!("kevent_ts error: {}", value);
```
This probably needs to be improved. Panicking the process will cause jails to linger around so we'd need to handle the recovering process before panicking.

Freaky commented:
I count 14 other places `kevent_ts` is called, some of which ignore the return value entirely.

It should probably be pulled out into a wrapper to handle EINTR in one place?

daniloegea commented:
Yeah I'm trying to go a step further here. I updated nix from 0.25.0 to 0.27.1 and it broke a lot of stuff. I replaced all the deprecated Kqueue code with the new API and fixed a bunch of other things that broke. I've got it compiling but xcd is not working properly. I'm still looking at it.

elliptic-kitty commented:
I have also upgraded nix to 0.27.1, xcd seems working fine.
This PR also added handling to EINTR

daniloegea commented:
Closing this PR in favor of https://github.com/michael-yuji/xc/pull/7
